Source: MontaVista Software, Inc.
Description:
  Reverse the order of iWMMXt register restores when using -mno-apcs-frame.

2003-11-04  Daniel Jacobowitz  <drow@mvista.com>

	* config/arm/arm.c (arm_output_epilogue): Reverse order of wldrd
	instructions if using a frame pointer.
	(arm_expand_prologue): Reverse order of PRE_DEC insns when saving
	iwmmxt registers.

--- gcc-3.3.1/gcc/config/arm/arm.c.orig	2003-11-04 12:17:16.000000000 -0500
+++ gcc-3.3.1/gcc/config/arm/arm.c	2003-11-04 12:19:21.000000000 -0500
@@ -7985,7 +7985,7 @@ arm_output_epilogue (really_return)
 	     between $fp and the register save area.  */
 	  lrm_count += current_function_pretend_args_size / 4;
 
-	  for (reg = FIRST_IWMMXT_REGNUM; reg <= LAST_IWMMXT_REGNUM; reg++)
+	  for (reg = LAST_IWMMXT_REGNUM; reg >= FIRST_IWMMXT_REGNUM; reg--)
 	    if (regs_ever_live[reg] && !call_used_regs[reg])
 	      {
 		asm_fprintf (f, "\twldrd\t%r, [%r, #-%d]\n", 
@@ -8823,7 +8823,7 @@ arm_expand_prologue ()
     }
 
   if (TARGET_IWMMXT)
-    for (reg = FIRST_IWMMXT_REGNUM; reg <= LAST_IWMMXT_REGNUM; reg++)
+    for (reg = LAST_IWMMXT_REGNUM; reg >= FIRST_IWMMXT_REGNUM; reg--)
       if (regs_ever_live[reg] && ! call_used_regs [reg])
 	{
 	  insn = gen_rtx_PRE_DEC (V2SImode, stack_pointer_rtx);
